engine/spec/components/citizens_advice_components/table_spec.rb

```diff
@@ -1,55 +1,79 @@
 # frozen_string_literal: true
 
 RSpec.describe CitizensAdviceComponents::Table, type: :component do
-  subject(:component) do
-    render_inline(
-      described_class.new(
-        header: header,
-        rows: rows,
-        caption: caption.presence
-      )
-    )
+  let(:sample_header) do
+    ["Your location", "Post box collection times"]
   end
 
-  let(:header) do
+  let(:sample_rows) do
     [
-      "Your location",
-      "Post box collection times"
+      ["City or town", "9am to 6.30pm"],
+      ["Areas with lots of businesses - known as commercial", "9am to 7.30pm"],
+      ["Very rural areas - for example, where there aren't many people", "9am to 4pm"],
+      ["Rest of the UK", "9am to 5.30pm"]
     ]
   end
 
-  let(:rows) do
-    [
-      [
-        "City or town",
-        "9am to 6.30pm"
-      ],
-      [
-        "Areas with lots of businesses - known as commercial",
-        "9am to 7.30pm"
-      ],
-      [
-        "Very rural areas - for example, where there aren't many people",
-        "9am to 4pm"
-      ],
-      [
-        "Rest of the UK",
-        "9am to 5.30pm"
-      ]
-    ]
+  context "when missing headers" do
+    subject(:component) do
+      described_class.new(header: [], rows: sample_rows)
+    end
+
+    it "does not render" do
+      render_inline component
+      expect(rendered_component).to have_no_table
+    end
+  end
+
+  context "when missing rows" do
+    subject(:component) do
+      described_class.new(header: sample_header, rows: [])
+    end
+
+    it "does not render" do
+      render_inline component
+      expect(rendered_component).to have_no_table
+    end
+  end
+
+  context "when valid table" do
+    subject(:component) do
+      described_class.new(
+        header: sample_header,
+        rows: sample_rows
+      )
+    end
+
+    it "matches snapshot" do
+      expect(render_inline(component).to_html).to match_snapshot("table")
+    end
   end
 
-  let(:caption) { "I am a nice table" }
+  context "when a caption is present" do
+    subject(:component) do
+      described_class.new(
+        header: sample_header,
+        rows: sample_rows,
+        caption: "Example caption"
+      )
+    end
 
-  it "renders correctly" do
-    expect(component.to_html).to match_snapshot("table")
+    it "renders a caption" do
+      render_inline component
+      expect(rendered_component).to have_css "caption", text: "Example caption"
```

Comment:
Not a problem, but just worth pointing out because capybara uses xpath under the hood, you can just use any of the capybara query matchers isolated. (So long as you're happy to). So if you wanted you could just then composite this to

`expect(rendered_component).to have_text "Example caption"` as well as increasing the amount so for instance

`expect(rendered_component).to have_css "caption", text: "Example caption", count: 1`

Reply:
<3